File: Cell-Model-Solver-v2/src/noble.cpp
#include "../include/noble.h"

void Noble::initConst (double CONSTANTS[], double RATES[], double STATES[])
{
    STATES[0] = -87;
    STATES[1] = 0.01;
    STATES[2] = 0.8;
    STATES[3] = 0.01;
    CONSTANTS[0] = 12;
    CONSTANTS[1] = 400000;
    CONSTANTS[2] = 40;
    CONSTANTS[3] = 75;
    CONSTANTS[4] = -60;
}

void Noble::compVariables (double t, double CONSTANTS[], double RATES[], double STATES[], double ALGEBRAIC[])
{
    RATES[0] = (- (ALGEBRAIC[4]+ALGEBRAIC[10]+ALGEBRAIC[11])/CONSTANTS[0]) * 1.0E-03;
    RATES[1] =  (ALGEBRAIC[1]*(1.00000 - STATES[1]) -  ALGEBRAIC[5]*STATES[1]) * 1.0E-03;
    RATES[2] =  (ALGEBRAIC[2]*(1.00000 - STATES[2]) -  ALGEBRAIC[6]*STATES[2]) * 1.0E-03;
    RATES[3] =  (ALGEBRAIC[3]*(1.00000 - STATES[3]) -  ALGEBRAIC[7]*STATES[3]) * 1.0E-03;
}

void Noble::compRates (double t, double CONSTANTS[], double RATES[], double STATES[], double ALGEBRAIC[])
{
    ALGEBRAIC[0] =  pow(STATES[1], 3.00000)*STATES[2]*CONSTANTS[1];
    ALGEBRAIC[1] = ( 100.000*(- STATES[0] - 48.0000))/(exp((- STATES[0] - 48.0000)/15.0000) - 1.00000);
    ALGEBRAIC[2] =  170.000*exp((- STATES[0] - 90.0000)/20.0000);
    ALGEBRAIC[3] = ( 0.100000*(- STATES[0] - 50.0000))/(exp((- STATES[0] - 50.0000)/10.0000) - 1.00000);
    ALGEBRAIC[4] =  (ALGEBRAIC[0]+140.000)*(STATES[0] - CONSTANTS[2]);
    ALGEBRAIC[5] = ( 120.000*(STATES[0]+8.00000))/(exp((STATES[0]+8.00000)/5.00000) - 1.00000);
    ALGEBRAIC[6] = 1000.00/(1.00000+exp((- STATES[0] - 42.0000)/10.0000));
    ALGEBRAIC[7] =  2.00000*exp((- STATES[0] - 90.0000)/80.0000);
    ALGEBRAIC[8] =  1200.00*exp((- STATES[0] - 90.0000)/50.0000)+ 15.0000*exp((STATES[0]+90.0000)/60.0000);
    ALGEBRAIC[9] =  1200.00*pow(STATES[3], 4.00000);
    ALGEBRAIC[10] =  (ALGEBRAIC[8]+ALGEBRAIC[9])*(STATES[0]+100.000);
    ALGEBRAIC[11] =  CONSTANTS[3]*(STATES[0] - CONSTANTS[4]);
}

void Noble::compRates_FE (double t, double CONSTANTS[], double RATES[], double STATES[], double ALGEBRAIC[])
{
    // V
    RATES[0] = (- (ALGEBRAIC[4]+ALGEBRAIC[10]+ALGEBRAIC[11])/CONSTANTS[0]) * 1.0E-03;
}

void Noble::RL (double DT, double CONSTANTS[], double RATES[], double STATES[], double ALGEBRAIC[])
{
    // Forward Euler for V
    STATES[0] = STATES[0] + RATES[0]*DT;
    
    // Rush-Larsen for m, h, n
    // m_inf = alpha_m / (alpha_m + beta_m)
    double m_inf = ALGEBRAIC[1] / (ALGEBRAIC[1] + ALGEBRAIC[5]);
    // tau_m = 1.0 / (alpha_m + beta_m)
    double tau_m = 1.0 / (ALGEBRAIC[1] + ALGEBRAIC[5]);
    STATES[1] = m_inf + ((STATES[1] - m_inf)*exp(-DT * 1.0E-03/tau_m));

    // h_inf = alpha_h / (alpha_h + beta_h)
    double h_inf = ALGEBRAIC[2] / (ALGEBRAIC[2] + ALGEBRAIC[6]);
    // tau_h = 1.0 / (alpha_h + beta_h)
    double tau_h = 1.0 / (ALGEBRAIC[2] + ALGEBRAIC[6]);
    STATES[2] = h_inf + ((STATES[2] - h_inf)*exp(-DT * 1.0E-03/tau_h));

    // n_inf = alpha_n / (alpha_n + beta_n)
    double n_inf = ALGEBRAIC[3] / (ALGEBRAIC[3] + ALGEBRAIC[7]);
    // tau_n = 1.0 / (alpha_n + beta_n)
    double tau_n = 1.0 / (ALGEBRAIC[3] + ALGEBRAIC[7]);
    STATES[3] = n_inf + ((STATES[3] - n_inf)*exp(-DT * 1.0E-03/tau_n));
}

/*
   There are a total of 12 entries in the algebraic variable array.
   There are a total of 4 entries in each of the rate and state variable arrays.
   There are a total of 5 entries in the constant variable array.
*/

/*
 * VOI is time in component environment (milisecond).
 * STATES[0] is V in component membrane (millivolt).
 * ALGEBRAIC[0] is g_Na in component sodium_channel (microS).
 * ALGEBRAIC[1] is alpha_m in component sodium_channel_m_gate (per_milisecond).
 * ALGEBRAIC[2] is alpha_h in component sodium_channel_h_gate (per_milisecond).
 * ALGEBRAIC[3] is alpha_n in component potassium_channel_n_gate (per_milisecond).
 * ALGEBRAIC[4] is i_Na in component sodium_channel (nanoA).
 * ALGEBRAIC[5] is beta_m in component sodium_channel_m_gate (per_milisecond).
 * ALGEBRAIC[6] is beta_h in component sodium_channel_h_gate (per_milisecond).
 * ALGEBRAIC[7] is beta_n in component potassium_channel_n_gate (per_milisecond).
 * ALGEBRAIC[8] is g_K1 in component potassium_channel (microS).
 * ALGEBRAIC[9] is g_K2 in component potassium_channel (microS).
 * ALGEBRAIC[10] is i_K in component potassium_channel (nanoA).
 * ALGEBRAIC[11] is i_Leak in component leakage_current (nanoA).
 * STATES[1] is m in component sodium_channel_m_gate (dimensionless).
 * STATES[2] is h in component sodium_channel_h_gate (dimensionless).
 * STATES[3] is n in component potassium_channel_n_gate (dimensionless).
 * RATES[0] is d/dt V in component membrane (millivolt).
 * RATES[1] is d/dt m in component sodium_channel_m_gate (dimensionless).
 * RATES[2] is d/dt h in component sodium_channel_h_gate (dimensionless).
 * RATES[3] is d/dt n in component potassium_channel_n_gate (dimensionless).
 * CONSTANTS[0] is Cm in component membrane (microF).
 * CONSTANTS[1] is g_Na_max in component sodium_channel (microS).
 * CONSTANTS[2] is E_Na in component sodium_channel (millivolt).
 * CONSTANTS[3] is g_L in component leakage_current (microS).
 * CONSTANTS[4] is E_L in component leakage_current (millivolt).
 * 
 */

File: Noble/CellML/noble.py
# Size of variable arrays:
sizeAlgebraic = 12
sizeStates = 4
sizeConstants = 5
from math import *
from numpy import *

def createLegends():
    legend_states = [""] * sizeStates
    legend_rates = [""] * sizeStates
    legend_algebraic = [""] * sizeAlgebraic
    legend_voi = ""
    legend_constants = [""] * sizeConstants
    legend_voi = "time in component environment (second)"
    legend_states[0] = "V in component membrane (millivolt)"
    legend_constants[0] = "Cm in component membrane (microF)"
    legend_algebraic[4] = "i_Na in component sodium_channel (nanoA)"
    legend_algebraic[10] = "i_K in component potassium_channel (nanoA)"
    legend_algebraic[11] = "i_Leak in component leakage_current (nanoA)"
    legend_constants[1] = "g_Na_max in component sodium_channel (microS)"
    legend_algebraic[0] = "g_Na in component sodium_channel (microS)"
    legend_constants[2] = "E_Na in component sodium_channel (millivolt)"
    legend_states[1] = "m in component sodium_channel_m_gate (dimensionless)"
    legend_states[2] = "h in component sodium_channel_h_gate (dimensionless)"
    legend_algebraic[1] = "alpha_m in component sodium_channel_m_gate (per_second)"
    legend_algebraic[5] = "beta_m in component sodium_channel_m_gate (per_second)"
    legend_algebraic[2] = "alpha_h in component sodium_channel_h_gate (per_second)"
    legend_algebraic[6] = "beta_h in component sodium_channel_h_gate (per_second)"
    legend_algebraic[8] = "g_K1 in component potassium_channel (microS)"
    legend_algebraic[9] = "g_K2 in component potassium_channel (microS)"
    legend_states[3] = "n in component potassium_channel_n_gate (dimensionless)"
    legend_algebraic[3] = "alpha_n in component potassium_channel_n_gate (per_second)"
    legend_algebraic[7] = "beta_n in component potassium_channel_n_gate (per_second)"
    legend_constants[3] = "g_L in component leakage_current (microS)"
    legend_constants[4] = "E_L in component leakage_current (millivolt)"
    legend_rates[0] = "d/dt V in component membrane (millivolt)"
    legend_rates[1] = "d/dt m in component sodium_channel_m_gate (dimensionless)"
    legend_rates[2] = "d/dt h in component sodium_channel_h_gate (dimensionless)"
    legend_rates[3] = "d/dt n in component potassium_channel_n_gate (dimensionless)"
    return (legend_states, legend_algebraic, legend_voi, legend_constants)

def initConsts():
    constants = [0.0] * sizeConstants; states = [0.0] * sizeStates;
    states[0] = -87
    constants[0] = 12
    constants[1] = 400000
    constants[2] = 40
    states[1] = 0.01
    states[2] = 0.8
    states[3] = 0.01
    constants[3] = 75
    constants[4] = -60
    return (states, constants)

def computeRates(voi, states, constants):
    rates = [0.0] * sizeStates; algebraic = [0.0] * sizeAlgebraic
    algebraic[1] = (100.000*(-states[0]-48.0000))/(exp((-states[0]-48.0000)/15.0000)-1.00000)
    algebraic[5] = (120.000*(states[0]+8.00000))/(exp((states[0]+8.00000)/5.00000)-1.00000)
    rates[1] = algebraic[1]*(1.00000-states[1])-algebraic[5]*states[1]
    algebraic[2] = 170.000*exp((-states[0]-90.0000)/20.0000)
    algebraic[6] = 1000.00/(1.00000+exp((-states[0]-42.0000)/10.0000))
    rates[2] = algebraic[2]*(1.00000-states[2])-algebraic[6]*states[2]
    algebraic[3] = (0.100000*(-states[0]-50.0000))/(exp((-states[0]-50.0000)/10.0000)-1.00000)
    algebraic[7] = 2.00000*exp((-states[0]-90.0000)/80.0000)
    rates[3] = algebraic[3]*(1.00000-states[3])-algebraic[7]*states[3]
    algebraic[0] = (power(states[1], 3.00000))*states[2]*constants[1]
    algebraic[4] = (algebraic[0]+140.000)*(states[0]-constants[2])
    algebraic[8] = 1200.00*exp((-states[0]-90.0000)/50.0000)+15.0000*exp((states[0]+90.0000)/60.0000)
    algebraic[9] = 1200.00*(power(states[3], 4.00000))
    algebraic[10] = (algebraic[8]+algebraic[9])*(states[0]+100.000)
    algebraic[11] = constants[3]*(states[0]-constants[4])
    rates[0] = -(algebraic[4]+algebraic[10]+algebraic[11])/constants[0]
    return(rates)

def computeAlgebraic(constants, states, voi):
    algebraic = array([[0.0] * len(voi)] * sizeAlgebraic)
    states = array(states)
    voi = array(voi)
    algebraic[1] = (100.000*(-states[0]-48.0000))/(exp((-states[0]-48.0000)/15.0000)-1.00000)
    algebraic[5] = (120.000*(states[0]+8.00000))/(exp((states[0]+8.00000)/5.00000)-1.00000)
    algebraic[2] = 170.000*exp((-states[0]-90.0000)/20.0000)
    algebraic[6] = 1000.00/(1.00000+exp((-states[0]-42.0000)/10.0000))
    algebraic[3] = (0.100000*(-states[0]-50.0000))/(exp((-states[0]-50.0000)/10.0000)-1.00000)
    algebraic[7] = 2.00000*exp((-states[0]-90.0000)/80.0000)
    algebraic[0] = (power(states[1], 3.00000))*states[2]*constants[1]
    algebraic[4] = (algebraic[0]+140.000)*(states[0]-constants[2])
    algebraic[8] = 1200.00*exp((-states[0]-90.0000)/50.0000)+15.0000*exp((states[0]+90.0000)/60.0000)
    algebraic[9] = 1200.00*(power(states[3], 4.00000))
    algebraic[10] = (algebraic[8]+algebraic[9])*(states[0]+100.000)
    algebraic[11] = constants[3]*(states[0]-constants[4])
    return algebraic

def solve_model():
    """Solve model with ODE solver"""
    from scipy.integrate import ode
    # Initialise constants and state variables
    (init_states, constants) = initConsts()

    # Set timespan to solve over
    voi = linspace(0,2,1000)

    # Construct ODE object to solve
    r = ode(computeRates)
    r.set_integrator('vode', method='bdf', atol=1e-06, rtol=1e-06, max_step=1)
    r.set_initial_value(init_states, voi[0])
    r.set_f_params(constants)

    # Solve model
    states = array([[0.0] * len(voi)] * sizeStates)
    states[:,0] = init_states
    for (i,t) in enumerate(voi[1:]):
        if r.successful():
            r.integrate(t)
            states[:,i+1] = r.y
        else:
            break

    # Compute algebraic variables
    algebraic = computeAlgebraic(constants, states, voi)
    return (voi, states, algebraic)

def plot_model(voi, states, algebraic):
    """Plot variables against variable of integration"""
    import pylab
    (legend_states, legend_algebraic, legend_voi, legend_constants) = createLegends()
    pylab.figure(1)
    #pylab.plot(voi,vstack((states,algebraic)).T)
    pylab.plot(voi,states[0])
    pylab.xlabel(legend_voi)
    #pylab.legend(legend_states + legend_algebraic, loc='best')
    pylab.show()

if __name__ == "__main__":
    (voi, states, algebraic) = solve_model()
    plot_model(voi, states, algebraic)